sc-poc/cmd/spacectl/commands/client/generate/golang/api.go

```diff
@@ -0,0 +1,259 @@
+package golang
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	"golang.org/x/tools/imports"
+)
+
+func GenerateAPI(spec *openapi3.T, pkgName string) (string, error) {
+	var b strings.Builder
+
+	// package name and imports
+	importsOut := generateImports(pkgName)
+	_, _ = b.WriteString(importsOut)
+
+	// client
+	clientOut := generateClient(spec)
+	_, _ = b.WriteString(clientOut)
+
+	// The generation code produces unindented horrors. Use the Go Imports
+	// to make it all pretty.
+	outBytes, err := imports.Process(pkgName+".go", []byte(b.String()), nil)
+	if err != nil {
+		return "", fmt.Errorf("error formatting Go code: %w", err)
+	}
+	return string(outBytes), nil
+}
+
+func generateImports(pkgName string) string {
+	s := fmt.Sprintf("package %s\n\n", pkgName)
+
+	imports := []string{"context", "net/http", "net/url", "bytes", "encoding/json"}
+	s += "import (\n"
+	for _, imp := range imports {
+		s += fmt.Sprintf("	%q\n", imp)
+	}
+	s += ")\n\n"
+	return s
+}
+
+func generateClient(doc *openapi3.T) string {
+	s := `
+// Client which conforms to the OpenAPI3 specification for SpaceCloud.
+type Client struct {
+	// The endpoint of the server. All the paths in
+	// the swagger spec will be appended to the server.
+	Server string
+
+	// Client for performing requests.
+	Client *http.Client
+}
+
+// Creates a new SpaceCloud Client
+func NewClient(server string) (*Client, error) {
+	// create a client with sane default values
+	client := Client{
+		Server: server,
+	}
+
+	// create httpClient, if not already present
+	if client.Client == nil {
+		client.Client = &http.Client{}
+	}
+	return &client, nil
+}`
+
+	// Generate operation methods
+	s += "\n\n"
+	for path, pathDef := range doc.Paths {
+		s += getFuncFromOperation(path, http.MethodGet, pathDef.Get)
+		s += getFuncFromOperation(path, http.MethodPost, pathDef.Post)
+		s += getFuncFromOperation(path, http.MethodPut, pathDef.Put)
+		s += getFuncFromOperation(path, http.MethodDelete, pathDef.Delete)
+	}
+	return s
+}
+
+func getFuncFromOperation(path, method string, operation *openapi3.Operation) string {
+	if !isOperationValidForTypeGen(operation) {
+		return ""
+	}
+
+	opName := getTypeName(operation.OperationID, false)
+	paramsOut := getOpParams(operation.Parameters)
+
+	paramsArg := ""
+	if paramsOut != "" {
+		paramsArg = "params " + opName + "Request"
+	}
+	var s string
+	s += fmt.Sprintf("// %s\n", opName)
+	switch method {
+	case "GET":
+		s += fmt.Sprintf("func (c *Client) %s(ctx context.Context, %s) (*%s, error) {\n", opName, paramsArg, opName+"Response")
```

Comment:
The response should not be the response object directly. instead it should be a struct having the following methods
- Raw() -> Returns http.Response (response.Body should still be readable. [Ref](https://stackoverflow.com/a/39792097)) 
- StatusCode() -> Returns the status code
- Result() -> Returns the struct you made if the status code was 200; This fn should only be available if openapi declared a 200 response code
- ResultXXX() -> Object for that status code (eg: `Result500`). struct name for these objects should be like `<OperationID>Response<Status>`